File: src/main/java/wraith/fabricaeexnihilo/modules/farming/TallPlantableItem.java
package wraith.fabricaeexnihilo.modules.farming;

import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.block.TallPlantBlock;
import net.minecraft.block.enums.DoubleBlockHalf;
import net.minecraft.item.Item;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.util.ActionResult;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TallPlantableItem extends Item {
    
    private final List<TallPlantBlock> plants;
    
    public TallPlantableItem(TallPlantBlock plant, FabricItemSettings settings) {
        this(Collections.singletonList(plant), settings);
    }
    
    public TallPlantableItem(List<TallPlantBlock> plants, FabricItemSettings settings) {
        super(settings);
        this.plants = plants;
    }
    
    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        var world = context.getWorld();
        var plantPos = context.getBlockPos().offset(context.getSide());
        var shuffledPlants = new ArrayList<>(plants);
        Collections.shuffle(shuffledPlants);
        for (var plant : shuffledPlants) {
            var lower = plant.getDefaultState().with(TallPlantBlock.HALF, DoubleBlockHalf.LOWER);
            var upper = plant.getDefaultState().with(TallPlantBlock.HALF, DoubleBlockHalf.UPPER);
            if (placementCheck(context) && lower.canPlaceAt(world, plantPos) && world.isAir(plantPos.up())) {
                world.setBlockState(plantPos, lower);
                world.setBlockState(plantPos.up(), upper);
                var player = context.getPlayer();
                if (player != null && !player.isCreative()) {
                    context.getStack().decrement(1);
                }
                return ActionResult.SUCCESS;
            }
        }
        return super.useOnBlock(context);
    }
    
    public boolean placementCheck(ItemUsageContext context) {
        return true;
    }
    
}

File: src/main/kotlin/exnihilofabrico/modules/farming/TallPlantableItem.kt
package exnihilofabrico.modules.farming

import exnihilofabrico.modules.base.BaseItem
import net.minecraft.block.TallPlantBlock
import net.minecraft.block.enums.DoubleBlockHalf
import net.minecraft.item.ItemUsageContext
import net.minecraft.util.ActionResult

class TallPlantableItem(val plants: List<TallPlantBlock>, settings: Settings): BaseItem(settings) {

    constructor(plant: TallPlantBlock, settings: Settings): this(listOf(plant), settings)

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        if(context.world.isClient)
            return ActionResult.SUCCESS
        val plantPos = context.blockPos.offset(context.side)
        for (plant in plants.shuffled()) {
            val lower = plant.defaultState.with(TallPlantBlock.HALF, DoubleBlockHalf.LOWER)
            val upper = plant.defaultState.with(TallPlantBlock.HALF, DoubleBlockHalf.UPPER)
            if(placementCheck(context) && lower.canPlaceAt(context.world, plantPos) && context.world.isAir(plantPos.up())){
                context.world.setBlockState(plantPos, lower)
                context.world.setBlockState(plantPos.up(), upper)
                if(context.player?.isCreative == false)
                    context.stack.decrement(1)
                return ActionResult.SUCCESS
            }
        }
        return super.useOnBlock(context)
    }

    fun placementCheck(context: ItemUsageContext): Boolean {
        return true
    }
}